flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveBulkFormatAdapter.java

```diff
@@ -0,0 +1,291 @@
+					DEFAULT_SIZE,
+					hiveVersion.startsWith("3"),
+					false
+			);
+		} else {
+			return new HiveMapRedBulkFormat();
+		}
+	}
+
+	private boolean useParquetVectorizedRead(HiveTablePartition partition) {
+		boolean isParquet = partition.getStorageDescriptor().getSerdeInfo().getSerializationLib()
+				.toLowerCase().contains("parquet");
+		if (!isParquet) {
+			return false;
+		}
+
+		for (int i : selectedFields) {
+			if (isVectorizationUnsupported(fieldTypes[i].getLogicalType())) {
+				LOG.info("Fallback to hadoop mapred reader, unsupported field type: " + fieldTypes[i]);
+				return false;
+			}
+		}
+
+		LOG.info("Use flink parquet ColumnarRowData reader.");
+		return true;
+	}
+
+	private static boolean isVectorizationUnsupported(LogicalType t) {
+		switch (t.getTypeRoot()) {
+			case CHAR:
+			case VARCHAR:
+			case BOOLEAN:
+			case BINARY:
+			case VARBINARY:
+			case DECIMAL:
+			case TINYINT:
+			case SMALLINT:
+			case INTEGER:
+			case BIGINT:
+			case FLOAT:
+			case DOUBLE:
+			case DATE:
+			case TIME_WITHOUT_TIME_ZONE:
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+			case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+				return false;
+			case TIMESTAMP_WITH_TIME_ZONE:
+			case INTERVAL_YEAR_MONTH:
+			case INTERVAL_DAY_TIME:
+			case ARRAY:
+			case MULTISET:
+			case MAP:
+			case ROW:
+			case DISTINCT_TYPE:
+			case STRUCTURED_TYPE:
+			case NULL:
+			case RAW:
+			case SYMBOL:
+			default:
+				return true;
+		}
+	}
+
+	private class HiveMapRedBulkFormat implements BulkFormat<RowData, HiveSourceSplit> {
+
+		private static final long serialVersionUID = 1L;
+
+		@Override
+		public Reader<RowData> createReader(Configuration config, HiveSourceSplit split)
+				throws IOException {
+			return new HiveReader(split);
+		}
+
+		@Override
+		public Reader<RowData> restoreReader(Configuration config, HiveSourceSplit split) throws IOException {
+			assert split.getReaderPosition().isPresent();
+			HiveReader hiveReader = new HiveReader(split);
+			hiveReader.seek(split.getReaderPosition().get().getRecordsAfterOffset());
+			return hiveReader;
+		}
+
+		@Override
+		public boolean isSplittable() {
+			return true;
+		}
+
+		@Override
+		public TypeInformation<RowData> getProducedType() {
+			return (TypeInformation<RowData>) TypeInfoDataTypeConverter.fromDataTypeToTypeInfo(producedDataType);
+		}
+	}
+
+	private class HiveReader implements BulkFormat.Reader<RowData> {
+
+		private final HiveMapredSplitReader hiveMapredSplitReader;
+		private final RowDataSerializer serializer;
+		private long numRead = 0;
+
+		private HiveReader(HiveSourceSplit split) throws IOException {
+			JobConf clonedConf = new JobConf(jobConfWrapper.conf());
+			addSchemaToConf(clonedConf);
+			HiveTableInputSplit oldSplit = new HiveTableInputSplit(-1, split.toMapRedSplit(), clonedConf, split.getHiveTablePartition());
+			hiveMapredSplitReader = new HiveMapredSplitReader(clonedConf, partitionKeys, fieldTypes, selectedFields, oldSplit, hiveShim);
```

Comment:
If you don't want migrate `HiveMapredSplitReader` now, can you create JIRA for this?

Reply:
Just created https://issues.apache.org/jira/browse/FLINK-19965 for that